core/trino-main/src/main/java/io/trino/sql/gen/columnar/DynamicPageFilter.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.gen.columnar;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import io.trino.Session;
+import io.trino.metadata.Metadata;
+import io.trino.operator.project.SelectedPositions;
+import io.trino.spi.Page;
+import io.trino.spi.connector.ColumnHandle;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.connector.DynamicFilter;
+import io.trino.spi.predicate.TupleDomain;
+import io.trino.spi.type.TypeManager;
+import io.trino.sql.PlannerContext;
+import io.trino.sql.ir.Expression;
+import io.trino.sql.ir.optimizer.IrExpressionOptimizer;
+import io.trino.sql.planner.DomainTranslator;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.relational.RowExpression;
+import it.unimi.dsi.fastutil.ints.IntOpenHashSet;
+import jakarta.annotation.Nullable;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Supplier;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static io.trino.sql.gen.columnar.FilterEvaluator.createColumnarFilterEvaluator;
+import static io.trino.sql.ir.optimizer.IrExpressionOptimizer.newOptimizer;
+import static io.trino.sql.relational.SqlToRowExpressionTranslator.translate;
+import static java.util.Objects.requireNonNull;
+
+public final class DynamicPageFilter
+{
+    private final Metadata metadata;
+    private final TypeManager typeManager;
+    private final Session session;
+    private final IrExpressionOptimizer irExpressionOptimizer;
+    private final DomainTranslator domainTranslator;
+    private final DynamicFilter dynamicFilter;
+    private final Map<ColumnHandle, Symbol> columnHandles;
+    private final Map<Symbol, Integer> sourceLayout;
+    private final double selectivityThreshold;
+
+    @Nullable
+    @GuardedBy("this")
+    private Supplier<FilterEvaluator> compiledDynamicFilter;
+    @Nullable
+    @GuardedBy("this")
+    private CompletableFuture<?> isBlocked;
```

Comment:
Why not to compare with previous `TupleDomain` for DF change instead? `io.trino.sql.planner.LocalDynamicFiltersCollector.TableSpecificDynamicFilter#currentPredicate` shouldn't change if DF didn't change so you could compare by identity if needed.

Alternatively you could use `io.trino.spi.connector.DynamicFilter#isComplete` instead of combination of `isBlocked/isAwaitable` (you don't really utilize blocking capabilities)

Alternatively you could similar approach as in https://github.com/trinodb/trino/pull/21888/files#diff-9ce339af9008637de326fb6d1be56265d7a0e1a55b9338d40264ba3eb93d280fR51:
```
    /**
     * Creates a {@code Supplier<StaticDynamicFilter>} that caches the {@link StaticDynamicFilter}.
     * Cache is reset whenever any underlying dynamic filter gets updated.
     */
    public static Supplier<StaticDynamicFilter> createStaticDynamicFilterSupplier(List<DynamicFilter> disjunctiveDynamicFilters)
    {
        AtomicReference<AtomicReference<StaticDynamicFilter>> dynamicFilterCache = new AtomicReference<>();
        resetDynamicFilterCache(dynamicFilterCache, disjunctiveDynamicFilters);
        return () -> {
            AtomicReference<StaticDynamicFilter> dynamicFilterReference = requireNonNull(dynamicFilterCache.get());
            StaticDynamicFilter dynamicFilter = dynamicFilterReference.get();
            if (dynamicFilter == null) {
                dynamicFilter = createStaticDynamicFilter(disjunctiveDynamicFilters);
                if (!dynamicFilterReference.compareAndSet(null, dynamicFilter)) {
                    return dynamicFilterReference.get();
                }
            }
            return dynamicFilter;
        };
    }

    private static void resetDynamicFilterCache(
            AtomicReference<AtomicReference<StaticDynamicFilter>> dynamicFilterCache,
            List<DynamicFilter> disjunctiveDynamicFilters)
    {
        if (areAwaitable(disjunctiveDynamicFilters)) {
            // reset dynamic filter cache whenever any underlying dynamic filter gets updated
            whenAnyUpdates(disjunctiveDynamicFilters).addListener(() -> resetDynamicFilterCache(dynamicFilterCache, disjunctiveDynamicFilters), directExecutor());
        }
        dynamicFilterCache.set(new AtomicReference<>());
    }

```
so that you recompile DF after you get notification